File: src/test/java/com/zepben/blobstore/sqlite/SqliteConnectionFactoryTest.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.blobstore.sqlite;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

import static com.zepben.blobstore.sqlite.SqliteConnectionFactory.SCHEMA_VERSION;
import static com.zepben.blobstore.sqlite.SqliteConnectionFactory.VERSION_TABLE;
import static com.zepben.testutils.exception.ExpectException.expect;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.containsString;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.core.Is.is;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class SqliteConnectionFactoryTest {

    private static final String tempDB = "store.db";
    private static Path dbFile;

    static {
        try {
            dbFile = Files.createTempFile(Path.of("/tmp"), tempDB, "");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @AfterAll
    public static void afterAll() throws IOException {
        Files.deleteIfExists(dbFile);
    }

    @BeforeEach
    public void before() throws IOException {
        Files.deleteIfExists(dbFile);
        dbFile = Files.createTempFile(Path.of("/tmp"), tempDB, "");
    }

    @Test
    @SuppressWarnings("SqlResolve")
    public void createsConnectionAndInitialisesDb() throws Exception {
        Set<String> tags = new HashSet<>(Arrays.asList("tag1", "tag2"));
        SqliteConnectionFactory factory = new SqliteConnectionFactory(dbFile, tags);
        try (Connection connection = factory.getConnection()) {
            assertNotNull(connection);

            try (Statement statement = connection.createStatement()) {
                ResultSet rs = statement.executeQuery("select * from " + VERSION_TABLE);
                assertTrue(rs.next());
                assertThat(rs.getString(1), equalTo(SCHEMA_VERSION));

                boolean isResultSet = statement.execute("select * from tag1");
                assertThat(isResultSet, is(true));

                isResultSet = statement.execute("select * from tag2");
                assertThat(isResultSet, is(true));

                isResultSet = statement.execute("select * from " + IdIndex.ID_INDEX_TABLE);
                assertThat(isResultSet, is(true));

                assertThat(connection.getMetaData().getURL(), containsString(dbFile.toString()));
            }
        }
    }

    @Test
    @SuppressWarnings("SqlResolve")
    public void failsOnVersionMismatch() throws Exception {
        Set<String> tags = new HashSet<>(Collections.singletonList("tag1"));
        SqliteConnectionFactory factory = new SqliteConnectionFactory(dbFile, tags);
        try (Connection connection = factory.getConnection()) {
            assertNotNull(connection);
            try (Statement statement = connection.createStatement()) {
                statement.executeUpdate("update " + VERSION_TABLE + " set version = 'foo'");
            }
        }

        expect(() -> new SqliteConnectionFactory(dbFile, tags).getConnection())
                .toThrow(SQLException.class)
                .withMessage("Failed to initialise sqlite database '" + dbFile.toString() + "'");

    }

}

File: src/test/kotlin/com/zepben/blobstore/sqlite/SqliteConnectionFactoryTest.kt
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.blobstore.sqlite

import com.zepben.testutils.exception.ExpectException.Companion.expect
import io.mockk.mockk
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.*
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test
import java.nio.file.Files
import java.nio.file.Path
import java.sql.SQLException

class SqliteConnectionFactoryTest {

    private var dbFile = Files.createTempFile(Path.of(System.getProperty("java.io.tmpdir")), "store.db", "")

    @AfterEach
    fun afterEach() {
        Files.deleteIfExists(dbFile)
    }

    @Test
    fun createsConnectionAndInitialisesDb() {
        val tags = setOf("tag1", "tag2")
        val factory = SqliteConnectionFactory(dbFile, tags)
        factory.getConnection().use { connection ->
            assertThat(connection, notNullValue())
            connection.createStatement().use { statement ->
                statement.executeQuery("select * from " + SqliteConnectionFactory.VERSION_TABLE).use { rs ->
                    assertThat("should have read the version record", rs.next())
                    assertThat(rs.getString(1), equalTo(SqliteConnectionFactory.SCHEMA_VERSION))
                }

                assertThat("should have created the table", statement.execute("select * from tag1"))
                assertThat("should have created the table", statement.execute("select * from tag2"))
                assertThat("should have created the table", statement.execute("select * from " + IdIndex.ID_INDEX_TABLE))
                assertThat(connection.metaData.url, containsString(dbFile.toString()))
            }
        }
    }

    @Test
    fun failsOnVersionMismatch() {
        val tags = setOf("tag1")
        val factory = SqliteConnectionFactory(dbFile, tags)
        factory.getConnection().use { connection ->
            assertThat(connection, notNullValue())
            connection.createStatement().use { statement ->
                statement.executeUpdate("update " + SqliteConnectionFactory.VERSION_TABLE + " set version = 'foo'")
            }
        }

        expect { SqliteConnectionFactory(dbFile, tags).getConnection() }
            .toThrow<SQLException>()
            .withMessage("Failed to initialise sqlite database '$dbFile'")
            .withCause<SQLException>()
    }

    @Test
    fun `can't use reserved names for tags`() {
        expect { SqliteConnectionFactory(mockk(), setOf(IdIndex.ID_INDEX_TABLE)) }
            .toThrow<IllegalArgumentException>()
            .withMessage("unsupported tag: ${IdIndex.ID_INDEX_TABLE}")

        expect { SqliteConnectionFactory(mockk(), setOf(SqliteConnectionFactory.VERSION_TABLE)) }
            .toThrow<IllegalArgumentException>()
            .withMessage("unsupported tag: ${SqliteConnectionFactory.VERSION_TABLE}")

        expect { SqliteConnectionFactory(mockk(), setOf("")) }
            .toThrow<IllegalArgumentException>()
            .withMessage("tags must not be empty strings")
    }

}
